src/devices/src/virtio/block/request.rs

```diff
@@ -183,22 +183,27 @@ impl Request {
         disk: &mut DiskProperties,
         mem: &GuestMemoryMmap,
     ) -> result::Result<u32, ExecuteError> {
-        let mut top: u64 = u64::from(self.data_len) / SECTOR_SIZE;
-        if u64::from(self.data_len) % SECTOR_SIZE != 0 {
-            top += 1;
-        }
-        top = top
-            .checked_add(self.sector)
-            .ok_or(ExecuteError::BadRequest(Error::InvalidOffset))?;
-        if top > disk.nsectors() {
-            return Err(ExecuteError::BadRequest(Error::InvalidOffset));
+        if self.request_type == RequestType::In || self.request_type == RequestType::Out {
+            // Check that the data length is a multiple of 512 as specified in the virtio standard.
+            // We could also check this at request parsing level.
+            if u64::from(self.data_len) & (SECTOR_SIZE - 1) != 0 {
```

Comment:
In theory `SECTOR_SIZE - 1 == b11111111` and any number that is not divisible with `512 (==b1000000000)` should be caught here since it must have one of the lower 8 bits set. I tested and `513_u64 & (SECTOR_SIZE - 1) == 1`

Reply:
Sorry, I misread this as `SECTOR_SHIFT`, which is 9 😆